enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+The following steps will ensure that kube-api is available post disaster recovery:
+
+- Dnsmasq service and Keepalived static pods will start running on master nodes
+- API-VIP address will be assigned to one of the master nodes by Keepalived
+- Dnsmasq should resolve api-int hostname to API-VIP address
+
+The above steps should be applied also for cluster shutdown and resume use-case.
+
+#### Timelined events: node shutdown, node restart, disaster recovery
+
+As the cluster-hosted-net-services-operator allows admins to use their own external Load Balancing and DNS resolution instead of the self-hosted stack, this section illustrates the events (in chronological order) for both self-hosted and external LB cases.
+
+##### Shutdown of master node (x) holds the VIP - self-hosted LB
+
+- T0: kube-apiserver reports failure on /readyz
+- T1: (= ~T0) Keepalived pod shutdown, Keepalived sends a priority 0 message to trigger VIP switchover.
+- T2: (=T1+0.6Sec) another master takes control on VIP, HAProxy on this node distributes API traffic to healthy kube-api backends (i.e: nodes != master-x)
+
+##### Shutdown of Master node (x) not holding the VIP - self-hosted LB
+
+- T0: kube-apiserver reports failure on /readyz
+- T1: (<= T0 + 2Sec) HAProxy (on the node holds the VIP) will mark master-x backend as failed, and should send new connections towards nodes != master-x.
+- T2: (= T0 + kube-api graceful termination timeout=90Sec ) kube-apiserver,Keepalived and HAProxy pods shutdown
+- T3: node shutdown
+
+##### Shutdown of Master node  - external LB,DNS
+
+For the external LB case, the VIP managed by Keepalived is not relevant.
+
+- T0: kube-apiserver reports failure on /readyz
+- T1: (= T0 + Y Sec , Y=external LB backend health detection time) external LB should mark master-x backend as failed, and should send new connections towards nodes != master-x.
+- T2: (= T0 + kube-api graceful termination timeout=90Sec ) kube-apiserver shutdown
+- T3: node shutdown
+
+##### Restart of Master node (x) - self-hosted LB
+
+As master node x being restarted the VIP is owned by another node.
+
+- T0: kube-apiserver returns OK on /readyz
+- T1: Keepalived static pod starts running, VIP should remain on node != master-x
+- T2: (<= T0 + 2Sec) HAProxy (on the node holds the VIP) will mark master-x backend as healthy, and should send new connections also towards master-x.
+- T3: (= T2 + 30Sec ) HAProxy pod starts running on master-x, that shouldn't affect API traffic since VIP remains on node != master-x
+
+##### Restart of Master node - external LB,DNS
+
+- T0: kube-apiserver returns OK on /readyz
+- T1: (= T0 + Y Sec , Y=external LB backend health detection time) external LB should mark master-x backend as ready, and should send new connections also towards master-x.
+
+##### Disaster recovery - self-hosted LB
+
+- T0: kube-apiserver recovered
+- T1: Keepalived static starts running, since this master is the only node running Keepalived at this point, it will hold the VIP.
```

Comment:
why is it the only one? Is it?

Reply:
Does it have to be the only one? If yes, this must have influence on the disaster recovery procedure I guess.

Reply:
AFAIK, at disaster recovery, you first recover a single master node and after that add the rest of the masters.
Anyhow, since Keepalived will continue to run as static pod, the VIP should be available even if more than one master node will be recovered.

The only difference here comparing to the current solution (based on static pods)  is when HAProxy DaemonSet pod will start running (and redirect traffic to LB) we'll have a connections break.

I don't think that it's critical if the disaster recovery will last another minute